### DISCUSSION TITLE
datasets/segment/coco8-seg/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#13252] [Docs]
# datasets/segment/coco8-seg/

Discover the COCO8-Seg: a compact but versatile instance segmentation dataset ideal for testing Ultralytics YOLOv8 detection approaches. Complete usage guide included.

https://docs.ultralytics.com/datasets/segment/coco8-seg/

**Comment by Mauricio566** [comment]
I really love their material. I have a question please help me. If the annotations of an image for a detection problem look like this... 1 0.5302913188934326 0.5031314492225647 0.2403690367937088 0.34829962253570557 So what does the annotation of a segmented image look like? because in the dataset I did not understand the annotation and you could help me understand this process. thank you

**Reply by glenn-jocher** [reply]
@Mauricio566 hello! Thanks for your kind words and for reaching out with your question. 🌟

For instance segmentation, unlike detection annotations which typically include class labels and bounding box coordinates (x_center, y_center, width, height), segmentation annotations are more complex. They include polygonal annotations or masks that precisely outline the object within an image.

In datasets like COCO, each segmented object in an image will have a corresponding mask defined by a set of polygon points. These points outline the exact shape of the object rather than just a bounding rectangle. This allows for a more accurate and detailed representation, especially useful for applications requiring precise object localization and shape delineation.

If you're working with the COCO8-Seg dataset, the annotations will include these detailed masks for each object in the images. You can view these annotations in the JSON format provided with the dataset, where each object's mask is represented as a series of coordinates outlining the object.

I hope this helps clarify the difference! If you have any more questions or need further assistance, feel free to ask. Happy coding! 😊

**Reply by jotar25** [reply]
this is actually not clear, should i separate masked color to multiple images? what is it the processs to  generate labels from multi mask image?

**Reply by glenn-jocher** [reply]
Hi @jotar25,

Thank you for your question! 😊 Let's clarify the process for generating labels from multi-mask images for instance segmentation.

In instance segmentation, each object in an image is annotated with a mask that precisely outlines its shape. These masks are typically represented as polygons or binary masks. Here's a step-by-step overview of the process:

1. **Separate Masks for Each Object**: Each object in the image should have its own mask. If you have a multi-mask image where different objects are represented by different colors, you will need to separate these into individual masks.

2. **Polygon Annotations**: For datasets like COCO, the masks are often stored as polygons. Each polygon is a list of coordinates that outline the object. These coordinates are stored in the dataset's annotation file (e.g., a JSON file for COCO).

3. **Binary Masks**: Alternatively, masks can be stored as binary images where the pixels belonging to an object are set to 1, and all other pixels are set to 0. Each object will have a separate binary mask.

4. **Annotation Format**: In the COCO format, each annotation includes:
   - `segmentation`: A list of polygons for the object.
   - `bbox`: The bounding box of the object.
   - `category_id`: The class of the object.
   - `image_id`: The ID of the image.

Here's an example of how a COCO annotation might look:
```json
{
  "segmentation": [[x1, y1, x2, y2, ..., xn, yn]],
  "bbox": [x, y, width, height],
  "category_id": 1,
  "image_id": 123
}
```

### Example Code for Generating Binary Masks
If you have a multi-mask image and need to generate individual binary masks, you can use the following Python code snippet:
```python
import cv2
import numpy as np

# Load the multi-mask image
multi_mask_image = cv2.imread('multi_mask_image.png')

# Assume each object is represented by a unique color
unique_colors = np.unique(multi_mask_image.reshape(-1, multi_mask_image.shape[2]), axis=0)

# Create binary masks for each unique color
binary_masks = []
for color in unique_colors:
    mask = cv2.inRange(multi_mask_image, color, color)
    binary_masks.append(mask)

# Save or process the binary masks
for i, mask in enumerate(binary_masks):
    cv2.imwrite(f'binary_mask_{i}.png', mask)
```

For more detailed information on the COCO8-Seg dataset and how to work with it, please refer to the [COCO8-Seg Dataset Documentation](https://docs.ultralytics.com/datasets/segment/coco8-seg/).

I hope this helps! If you have any more questions or need further assistance, feel free to ask. Happy coding! 🚀

**Comment by kh-diab** [comment]
Hello
I am struggling because I followed the same coco format in the yaml file but still get this error


      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
  0%|          | 0/2 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/ultralytics/utils/loss.py", line 289, in __call__
    targets = torch.cat((batch_idx, batch["cls"].view(-1, 1), batch["bboxes"]), 1)
RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 66 but got size 0 for tensor number 1 in the list.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/kdiab/Documents/OST/workspace/dataset/coding/Train_Model/train segment model V6.0.py", line 33, in <module>
    results =model.train(
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/ultralytics/engine/model.py", line 652, in train
    self.trainer.train()
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 204, in train
    self._do_train(world_size)
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 381, in _do_train
    self.loss, self.loss_items = self.model(batch)
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 101, in forward
    return self.loss(x, *args, **kwargs)
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 283, in loss
    return self.criterion(preds, batch)
  File "/home/kdiab/.conda/envs/yolov8/lib/python3.10/site-packages/ultralytics/utils/loss.py", line 294, in __call__
    raise TypeError(
TypeError: ERROR ❌ segment dataset incorrectly formatted or not a segment dataset.
This error can occur when incorrectly training a 'segment' model on a 'detect' dataset, i.e. 'yolo train model=yolov8n-seg.pt data=coco8.yaml'.
Verify your dataset is a correctly formatted 'segment' dataset using 'data=coco8-seg.yaml' as an example.


and this is my yaml file 


# Ultralytics YOLO , AGPL-3.0 license
# COCO8-seg dataset (first 8 images from COCO train2017) by Ultralytics
# Documentation: https://docs.ultralytics.com/datasets/segment/coco8-seg/
# Example usage: yolo train data=coco8-seg.yaml
# parent
# ├── ultralytics
# └── datasets
#     └── coco8-seg  ← downloads here (1 MB)

# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: /home/kdiab/Documents/OST/workspace/dataset/model_train/Group-C-tiny-Segment # dataset root dir
train: /home/kdiab/Documents/OST/workspace/dataset/model_train/Group-C-tiny-Segment/images/train # train images (relative to 'path') 4 images
val: /home/kdiab/Documents/OST/workspace/dataset/model_train/Group-C-tiny-Segment/images/val # val images (relative to 'path') 4 images
test: # test images (optional)


# Classes
names:
  0: weed
  1: rumex

and I am using model
model = YOLO('yolov8x-seg.pt')

my labeled files is as follows 

0 0.094970703125 0.46401098901098903 0.09033203125 0.160989010989011
0 0.07861328125 0.46153846153846156 0.0556640625 0.03296703296703297
0 0.116943359375 0.43763736263736264 0.02783203125 0.036813186813186814

can you please advise what is the error

**Reply by glenn-jocher** [reply]
@kh-diab it looks like the error you're encountering is due to a mismatch in the dataset format. Ensure that your dataset is correctly formatted for segmentation. Verify that your label files contain segmentation masks and not just bounding boxes. Also, make sure you are using the latest version of the Ultralytics package. If the issue persists, please check the dataset structure against the COCO8-Seg example provided in the documentation: https://docs.ultralytics.com/datasets/segment/coco8-seg/. This should help resolve the formatting issue.
